[dynamo] add initial dynamo support for DTensor

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103146
* #105308

This PR adds initial dynamo support for DTensor, in particular, it:
- allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.
- We use `allow_in_graph` to include `DTensor` and `DTensor.from_local` to be represented as `TorchVariable`
- The dtensor created becomes a normal `TensorVariable` and it would insert any tensor operations to the output graph just like torch.Tensor
- note that dtensor have a new instance method `redistribute` compare to plain tensor, and we currently special handle it in `TensorVariable`

`from_local` and `redistribute` both accepts some non-trival metadata as arguments (i.e. DeviceMesh, Placement) which fx.Graph does not support. In order to let these two APIs appear in the dynamo captured graph, we encoded the metadata into a new_function (like `functools.partial`) and the new function only accepts prim args (i.e. tensor), then we put `call_function` with this new_function to the graph. This is suggested by @ezyang. The underlying rationale here is that the metadata will not change across the graph invocations so it's safe to encode them.


Captured graph:
```
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:685, code: dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
        prim_from_local = torch__dynamo_variables_torch_prim_from_local(l_x_, run_check = False);  l_x_ = None
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:686, code: return dt.redistribute(mesh, [Replicate()]).to_local() + 2
        prim_redistribute = torch__dynamo_variables_tensor_prim_redistribute(prim_from_local);  prim_from_local = None
        to_local = prim_redistribute.to_local();  prim_redistribute = None
        add = to_local + 2;  to_local = None
        return (add,)
```

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

wanchaol committed:
[WIP][dynamo] add initial dynamo support for DTensor

This PR adds initial dynamo support for DTensor, in particular, it
allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.

The fakification of DTensor mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540 and would need to add some
principled approach for those `isinstance` checks

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it
allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.

The fakification of DTensor mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540 and would need to add some
principled approach for those `isinstance` checks

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

jbschlosser commented on torch/_dynamo/variables/builder.py:
```diff
@@ -161,7 +161,7 @@ def __post_init__(self):
             self._example = TensorWeakRef(self._example)
             assert isinstance(
-                self.fake_tensor, torch._subclasses.fake_tensor.FakeTensor
+                self.fake_tensor, (torch._subclasses.fake_tensor.FakeTensor, torch.distributed._tensor.DTensor)
```
In my nested tensor PR, I get around this by hacking the instancecheck magic method on FakeTensor:
```python
class FakeTensorMeta(torch._C._TensorMeta):
    def __instancecheck__(self, instance):
        from torch.nested._nested_tensor import NestedTensor

        if isinstance(instance, NestedTensor):
            # Check if NT's buffer is fake
            return isinstance(instance.buffer, self)

        return super().__instancecheck__(instance)
```

This has the benefit that the checks for `isinstance(x, FakeTensor)` throughout dynamo don't need to change.

This isn't particularly nice though, so a more principled solution along the lines of Brian's `is_fake()` would likely be better.

wanchaol replied:
Cool thanks for the suggestions! I'll give this a try

wconstab commented on test/distributed/_tensor/test_dtensor.py:
```diff
@@ -606,5 +607,33 @@ def test_split_tensor(self) -> None:
+        def fn(x):
+            return x.to_local()
+
+        x = DTensor.from_local(torch.rand(1), mesh, [Replicate()], run_check=False)
```
iiuc this test just simulates one rank and i am not sure if DTensor does anything to the data from the torch.rand tensor.  Do we need to at least test some comm operation to make sure dynamo tracing is working?

wanchaol replied:
Yep we would need some comm operation to make sure it's getting captured, I'll try to get the `redistribute` call also works with dynamo. What I am not sure yet is whether we expand the `redistribute` call in dynamo or we just make it a `call_method`, and expand that in the aot autograd, probably this would need to be done in aot autograd

wanchaol commented:
Alright, so the whole failure of CI is because lots of CI does not have distributed package built in... I'll switch those isinstance checks to `is_fake_tensor` fashion and hopefully all of them goes away

wanchaol committed:
Update on "[WIP][dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it
allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.

The fakification of DTensor mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540 and would need to add some
principled approach for those `isinstance` checks

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it
allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.

The fakification of DTensor mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540 and would need to add some
principled approach for those `isinstance` checks

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it
allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.

The fakification of DTensor mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540 and would need to add some
principled approach for those `isinstance` checks

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -590,6 +590,49 @@ def has_bool_key(v):
+            from .user_defined import UserDefinedObjectVariable
+
+            def as_value(v):
+                if isinstance(v, list):
+                    return [as_value(x) for x in v]
+                if isinstance(v, UserDefinedObjectVariable):
+                    from torch.distributed._tensor.placement_types import Shard
+
+                    placement_val = v.value
+                    # XXX: nasty hack in order to get the real object from UserDefinedObject with attributes
+                    if isinstance(
+                        placement_val, Shard
+                    ) and tx.output.side_effects.is_attribute_mutation(v):
+                        placement_val.dim = tx.output.side_effects.load_attr(
+                            v, "dim"
+                        ).value
+                    return placement_val
+                elif isinstance(v, variables.BaseListVariable):
+                    return v.python_type()([as_value(x) for x in v.items])
```
Why would there be a mix of VariableTracker and raw list types in this function? sus

wanchaol replied:
hmm yeah maybe we don't need the list check as we already have the BaseListVariable, removing that

wanchaol replied:
using top level VT and defined `as_python_constant` allow us to remove this as a whole, thanks for the suggestion!

voznesenskym commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -590,6 +590,49 @@ def has_bool_key(v):
+        elif name == "redistribute":
+            from .user_defined import UserDefinedObjectVariable
+
+            def as_value(v):
```
Hmm, I think as_python_constant might already do this? 

For baselist:
```
    def as_python_constant(self):
        return self.python_type()([x.as_python_constant() for x in self.items])
```

However, UserDefinedObjectVariable does not have an as_python_constant - if you want this generally, that should be its own PR. HOWEVER - if you only want it for `torch.distributed._tensor.placement_types.Shard` - you can make a DistributedTensorPlacementShardVariableTracker and encapsulate both this logic, and the `as_python_constant` without needing to cook up this generally recursive `as_value`. So, I guess the question is...do you need any UDO?

wanchaol replied:
good call, placement types currently be lowered as UDO, as that's why I need to do this side effect local attr reading in order to get the same object (which is kinda hacky). Maybe what I should do here is to add a new VT that tracks Placement types instead of using UDO.

wanchaol replied:
updated to use top level VT instead.

voznesenskym commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -590,6 +590,49 @@ def has_bool_key(v):
+                ),
+                **options,
+            )
+            return tensor_variable
```
nice, ultranit- just `return wrap_fx_proxy...`

voznesenskym commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -154,6 +155,16 @@ def _is_tensor_constructor(func: OpOverload):
+    if isinstance(x, FakeTensor):
+        return True
+    if is_traceable_wrapper_subclass(x):
+        flattened_tensors, _ = type(x).__tensor_flatten__(x)
+        # need to recurse because we could have nested subclasses
+        return all(is_fake(x) for x in flattened_tensors)
```
What is the invariant here? When is it possible for this to be partially fake?

bdhirsh replied:
Good question - we should probably assert here. Something's probably gone horribly wrong if we have a wrapper subclass that holds some real tensors and some fake tensors.

voznesenskym commented on torch/distributed/_tensor/_dynamo_utils.py:
```diff
@@ -0,0 +1,6 @@
+from torch.distributed._tensor.api import DTensor
+
+# dynamo/torch.compile utils for
+allow_in_graph(DTensor)
+allow_in_graph(DTensor.from_local)
```
nice :)

bdhirsh commented on torch/_subclasses/meta_utils.py:
```diff
@@ -551,6 +555,15 @@ def __call__(
+                        t, shape_env=shape_env, callback=callback, source=source
+                    ),
+                )
+                return out
```
Unfortunately I don't think this works :(. In particular, in my AOTDispatch prototype, I realized that fake-ifying this way causes your subclass to lose autograd metadata (the original `DTensor` has `requires_grad=True`, the updated one has `requires_grad=False`).

@wanchaol and I talked a bit about this offline. I have an updated version that seems to preserve autograd metadata in my prototype here: https://github.com/pytorch/pytorch/pull/104483/files#diff-ae6c64c5b6b4fbfa293a1ce7e82d4a9b9b7ee7cdca8ef66312031b52c69a0b4aR155. This version works on my basic examples, but fails for `DTensor` (because the meta conversion logic ends up calling `.set_()`, which `DTensor` sees and can't handle).

wanchaol replied:
Yeah we need to look into what this `meta_tensor` did for the `set_` call and see how that applies to DTensor. I once understood how that functions works but now my memory got damaged on it.. I'll re-read this and see what's going on there :)

wanchaol replied:
discussed offline, @bdhirsh is going to make this sound in https://github.com/pytorch/pytorch/pull/104482

wconstab commented:
Regarding capturing devicemesh/placement into a function and capturing the function-

This is a different approach than traceable collectives took, namely that TC has code that desugars types like devicemesh/processgroup into lists of ranks and a tag, which dynamo can then put into the graph as plain types.

I am open to this approach and wonder if we should use it for TC too.  Also, for TC there was some desire to encode only a 'tag' in the trace to make the trace more reusable across cluster sizes- but that may conflict with some usage by XLA.  cc @kumpera @JackCaoG

In any case, should we eventually align the approaches, or is there a good reason for them to be different?

wanchaol commented:
> Regarding capturing devicemesh/placement into a function and capturing the function-
> 
> This is a different approach than traceable collectives took, namely that TC has code that desugars types like devicemesh/processgroup into lists of ranks and a tag, which dynamo can then put into the graph as plain types.
> 
> I am open to this approach and wonder if we should use it for TC too. Also, for TC there was some desire to encode only a 'tag' in the trace to make the trace more reusable across cluster sizes- but that may conflict with some usage by XLA. cc @kumpera @JackCaoG
> 
> In any case, should we eventually align the approaches, or is there a good reason for them to be different?

@wconstab Yeah it's a good point. I was also considering the custom op approach (i.e. [these lines](https://github.com/pytorch/pytorch/pull/103146/files#diff-8b132984dbf1ba8ba0bfc128649d4e307137b171164579b6e20e4e9453380278R656-R668)) The custom op approach effectively works very similar to the TC approach, but it requires some non-trival rewrite for the current DTensor APIs (i.e. we need to map DeviceMesh to a "tensor/list" and caches the device mesh so that we can re-fetch it inside the custom op, and we also need to map the placement types (i.e. `Replicate()` and `Shard(dim)`) to some prim types (i don't know what's proper there, maybe simply integers?), then remap it back to placement types inside the custom op.

The non-trival refactoring is the reason I didn't go with it in dynamo. The aot_autograd pass later would trace the aten op that calls inside those two APIs (`from_local` and `redistribute`). I am ok with approaching this again in follow up PRs so that we stay aligned about how the ranks get lowered to aten op in both TC and DTensor. I think one thing that's not clear to me is that since dynamo works on higher level IR, should it makes the `allow_in_graph` works for abitrary data types like DeivceMesh/ProcessGroup? Maybe extending the fx IR capability to handle custom types is one way to go. cc @ezyang

wconstab commented:
Well I wasn't lobbying for one approach over the other just alignment or a reason to not align. Also don't think it's urgent to change in this PR.

ezyang commented:
As I discussed with @wanchaol in our 1:1, I think the long term right thing to do is to design some way for FX base argument types to be extended, with the requirement that such types must be value-like and be renderable into Python AST (there's also some shenanigans with import statements that also has to be dealt with.) In the short term, making your ops accept string reps for device mesh and co, and parsing them on the inside, is an easy workaround absent this mechanism.

voznesenskym commented:
> making your ops accept string reps for device mesh

We have prior art for this kind desugaring for process groups, lmk if you need a hand.

wanchaol committed:
Update on "[WIP][dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it:
- allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.
- We use `allow_in_graph` to include `DTensor` and `DTensor.from_local` to be represented as `TorchVariable`
- The dtensor created becomes a normal `TensorVariable` and it would insert any tensor operations to the output graph just like torch.Tensor
- note that dtensor have a new instance method `redistribute` compare to plain tensor, and we currently special handle it in `TensorVariable`

`from_local` and `redistribute` both accepts some non-trival metadata as arguments (i.e. DeviceMesh, Placement) which fx.Graph does not support. In order to let these two APIs appear in the dynamo captured graph, we encoded the metadata into a new_function (like `functools.partial`) and the new function only accepts prim args (i.e. tensor), then we put `call_function` with this new_function to the graph. This is suggested by ezyang. The underlying rationale here is that the metadata will not change across the graph invocations so it's safe to encode them.

The fakification of DTensor mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540 and would need to add some
principled approach for those `isinstance` checks


Captured graph:
```
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:685, code: dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
        prim_from_local = torch__dynamo_variables_torch_prim_from_local(l_x_, run_check = False);  l_x_ = None
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:686, code: return dt.redistribute(mesh, [Replicate()]).to_local() + 2
        prim_redistribute = torch__dynamo_variables_tensor_prim_redistribute(prim_from_local);  prim_from_local = None
        to_local = prim_redistribute.to_local();  prim_redistribute = None
        add = to_local + 2;  to_local = None
        return (add,)
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

wanchaol commented:
> > making your ops accept string reps for device mesh
> 
> We have prior art for this kind desugaring for process groups, lmk if you need a hand.

Thanks! Yeah I think we should try this out and see if we could make the metadata related APIs be custom ops, I think there're two missing parts we need to figure out:
- DeviceMesh need to have a global caching layer similar to the process group, so can we could desugar the mesh inside the op itself, this should not be very hard and I could give it a try in a follow up PR.
- Sharding strategy (replicate/shard) can be serialized to string then pass to the op, when desugaring it we can see if it could construct back the old instance (we might need to make them singleton if necessary).

wanchaol committed:
Update on "[WIP][dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it:
- allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.
- We use `allow_in_graph` to include `DTensor` and `DTensor.from_local` to be represented as `TorchVariable`
- The dtensor created becomes a normal `TensorVariable` and it would insert any tensor operations to the output graph just like torch.Tensor
- note that dtensor have a new instance method `redistribute` compare to plain tensor, and we currently special handle it in `TensorVariable`

`from_local` and `redistribute` both accepts some non-trival metadata as arguments (i.e. DeviceMesh, Placement) which fx.Graph does not support. In order to let these two APIs appear in the dynamo captured graph, we encoded the metadata into a new_function (like `functools.partial`) and the new function only accepts prim args (i.e. tensor), then we put `call_function` with this new_function to the graph. This is suggested by ezyang. The underlying rationale here is that the metadata will not change across the graph invocations so it's safe to encode them.

The fakification of DTensor mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540 and would need to add some
principled approach for those `isinstance` checks


Captured graph:
```
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:685, code: dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
        prim_from_local = torch__dynamo_variables_torch_prim_from_local(l_x_, run_check = False);  l_x_ = None
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:686, code: return dt.redistribute(mesh, [Replicate()]).to_local() + 2
        prim_redistribute = torch__dynamo_variables_tensor_prim_redistribute(prim_from_local);  prim_from_local = None
        to_local = prim_redistribute.to_local();  prim_redistribute = None
        add = to_local + 2;  to_local = None
        return (add,)
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it:
- allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.
- We use `allow_in_graph` to include `DTensor` and `DTensor.from_local` to be represented as `TorchVariable`
- The dtensor created becomes a normal `TensorVariable` and it would insert any tensor operations to the output graph just like torch.Tensor
- note that dtensor have a new instance method `redistribute` compare to plain tensor, and we currently special handle it in `TensorVariable`

`from_local` and `redistribute` both accepts some non-trival metadata as arguments (i.e. DeviceMesh, Placement) which fx.Graph does not support. In order to let these two APIs appear in the dynamo captured graph, we encoded the metadata into a new_function (like `functools.partial`) and the new function only accepts prim args (i.e. tensor), then we put `call_function` with this new_function to the graph. This is suggested by ezyang. The underlying rationale here is that the metadata will not change across the graph invocations so it's safe to encode them.

The fakification of DTensor mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540 and would need to add some
principled approach for those `isinstance` checks


Captured graph:
```
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:685, code: dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
        prim_from_local = torch__dynamo_variables_torch_prim_from_local(l_x_, run_check = False);  l_x_ = None
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:686, code: return dt.redistribute(mesh, [Replicate()]).to_local() + 2
        prim_redistribute = torch__dynamo_variables_tensor_prim_redistribute(prim_from_local);  prim_from_local = None
        to_local = prim_redistribute.to_local();  prim_redistribute = None
        add = to_local + 2;  to_local = None
        return (add,)
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

wanchaol commented:
This should be ready for another review :)

wanchaol committed:
Update on "[dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it:
- allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.
- We use `allow_in_graph` to include `DTensor` and `DTensor.from_local` to be represented as `TorchVariable`
- The dtensor created becomes a normal `TensorVariable` and it would insert any tensor operations to the output graph just like torch.Tensor
- note that dtensor have a new instance method `redistribute` compare to plain tensor, and we currently special handle it in `TensorVariable`

`from_local` and `redistribute` both accepts some non-trival metadata as arguments (i.e. DeviceMesh, Placement) which fx.Graph does not support. In order to let these two APIs appear in the dynamo captured graph, we encoded the metadata into a new_function (like `functools.partial`) and the new function only accepts prim args (i.e. tensor), then we put `call_function` with this new_function to the graph. This is suggested by ezyang. The underlying rationale here is that the metadata will not change across the graph invocations so it's safe to encode them.

The fakification of DTensor mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540 and would need to add some
principled approach for those `isinstance` checks


Captured graph:
```
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:685, code: dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
        prim_from_local = torch__dynamo_variables_torch_prim_from_local(l_x_, run_check = False);  l_x_ = None
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:686, code: return dt.redistribute(mesh, [Replicate()]).to_local() + 2
        prim_redistribute = torch__dynamo_variables_tensor_prim_redistribute(prim_from_local);  prim_from_local = None
        to_local = prim_redistribute.to_local();  prim_redistribute = None
        add = to_local + 2;  to_local = None
        return (add,)
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -1267,7 +1267,7 @@ def get_fake_value(node, tx):
     def fake_wrapper(e):
         if isinstance(e, torch.Tensor):
-            assert isinstance(e, FakeTensor)
+            assert is_fake(e)
```
if this PR is held up in code review, you should send in these changes on their own

wanchaol committed:
Update on "[dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it:
- allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.
- We use `allow_in_graph` to include `DTensor` and `DTensor.from_local` to be represented as `TorchVariable`
- The dtensor created becomes a normal `TensorVariable` and it would insert any tensor operations to the output graph just like torch.Tensor
- note that dtensor have a new instance method `redistribute` compare to plain tensor, and we currently special handle it in `TensorVariable`

`from_local` and `redistribute` both accepts some non-trival metadata as arguments (i.e. DeviceMesh, Placement) which fx.Graph does not support. In order to let these two APIs appear in the dynamo captured graph, we encoded the metadata into a new_function (like `functools.partial`) and the new function only accepts prim args (i.e. tensor), then we put `call_function` with this new_function to the graph. This is suggested by ezyang. The underlying rationale here is that the metadata will not change across the graph invocations so it's safe to encode them.

The fakification of DTensor mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540 and would need to add some
principled approach for those `isinstance` checks


Captured graph:
```
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:685, code: dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
        prim_from_local = torch__dynamo_variables_torch_prim_from_local(l_x_, run_check = False);  l_x_ = None
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:686, code: return dt.redistribute(mesh, [Replicate()]).to_local() + 2
        prim_redistribute = torch__dynamo_variables_tensor_prim_redistribute(prim_from_local);  prim_from_local = None
        to_local = prim_redistribute.to_local();  prim_redistribute = None
        add = to_local + 2;  to_local = None
        return (add,)
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

wanchaol committed:
Update on "[dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it:
- allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.
- We use `allow_in_graph` to include `DTensor` and `DTensor.from_local` to be represented as `TorchVariable`
- The dtensor created becomes a normal `TensorVariable` and it would insert any tensor operations to the output graph just like torch.Tensor
- note that dtensor have a new instance method `redistribute` compare to plain tensor, and we currently special handle it in `TensorVariable`

`from_local` and `redistribute` both accepts some non-trival metadata as arguments (i.e. DeviceMesh, Placement) which fx.Graph does not support. In order to let these two APIs appear in the dynamo captured graph, we encoded the metadata into a new_function (like `functools.partial`) and the new function only accepts prim args (i.e. tensor), then we put `call_function` with this new_function to the graph. This is suggested by ezyang. The underlying rationale here is that the metadata will not change across the graph invocations so it's safe to encode them.


Captured graph:
```
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:685, code: dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
        prim_from_local = torch__dynamo_variables_torch_prim_from_local(l_x_, run_check = False);  l_x_ = None
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:686, code: return dt.redistribute(mesh, [Replicate()]).to_local() + 2
        prim_redistribute = torch__dynamo_variables_tensor_prim_redistribute(prim_from_local);  prim_from_local = None
        to_local = prim_redistribute.to_local();  prim_redistribute = None
        add = to_local + 2;  to_local = None
        return (add,)
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

wanchaol committed:
Update on "[dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it:
- allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.
- We use `allow_in_graph` to include `DTensor` and `DTensor.from_local` to be represented as `TorchVariable`
- The dtensor created becomes a normal `TensorVariable` and it would insert any tensor operations to the output graph just like torch.Tensor
- note that dtensor have a new instance method `redistribute` compare to plain tensor, and we currently special handle it in `TensorVariable`

`from_local` and `redistribute` both accepts some non-trival metadata as arguments (i.e. DeviceMesh, Placement) which fx.Graph does not support. In order to let these two APIs appear in the dynamo captured graph, we encoded the metadata into a new_function (like `functools.partial`) and the new function only accepts prim args (i.e. tensor), then we put `call_function` with this new_function to the graph. This is suggested by ezyang. The underlying rationale here is that the metadata will not change across the graph invocations so it's safe to encode them.


Captured graph:
```
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:685, code: dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
        prim_from_local = torch__dynamo_variables_torch_prim_from_local(l_x_, run_check = False);  l_x_ = None
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:686, code: return dt.redistribute(mesh, [Replicate()]).to_local() + 2
        prim_redistribute = torch__dynamo_variables_tensor_prim_redistribute(prim_from_local);  prim_from_local = None
        to_local = prim_redistribute.to_local();  prim_redistribute = None
        add = to_local + 2;  to_local = None
        return (add,)
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -0,0 +1,113 @@
+
+            if type(value) is type and issubclass(value, Placement):
+                return True
+        return False
```
should this raise unimplemented? Maybe even during construction? Is there a valid case for `PlacementClassVariable` outside of distributed?

wanchaol replied:
I think we can't raise unimplemented here because the these `statemethod` `is_placement_type` `is_device_mesh` are used in the variable builder to check if the value passed in to create the VT is a certain type or not, if not, it would still need to go through other checks and see if it should create other VTs, i.e. if it's not a placement type, it might still need to create a UDO VT by default, so I feel we can not raise unimplemented for these type checks?

voznesenskym commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -0,0 +1,113 @@
+
+
+class PlacementClassVariable(VariableTracker):
+    def __init__(self, value, **kwargs):
+        super().__init__(**kwargs)
+        self.value = value
```
consider a class between these and `VariableTracker` - something like `DistributedVariableTracker` where init does `if not torch.distributed.is_available(): unimplemented(...)`

voznesenskym commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -0,0 +1,113 @@
+        # and __setattr__ methods, the latter is for case like `Shard(dim)`
+        if name in allowed_methods:
+            try:
+                method = inspect.getattr_static(type(self.value), name)
```
you need to prove there are no user defined getattrs, or this might invoke arbitrary user code.

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -634,6 +627,26 @@ def index_source(key):
+            return PlacementClassVariable(
+                value,
+                source=self.source,
+                guards=make_guards(GuardBuilder.FUNCTION_MATCH),
```
This is just ID_MATCH with a local scope. Use ID MATCH, or, better yet, write a custom guard for this type for the things you care about (Can be a TODO).

voznesenskym commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -591,6 +591,26 @@ def has_bool_key(v):
+                tx=tx,
+                proxy=tx.output.create_proxy(
+                    "call_function",
+                    redistribute_fn_with_prim_types,
```
Do you mind showing what a graph with this looks like? I am not certain you can just capture it like this and close over `args_as_value` soundly

wanchaol replied:
Sure here is the captured graph:

```python
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:685, code: dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
        prim_from_local = torch__dynamo_variables_torch_prim_from_local(l_x_, run_check = False);  l_x_ = None
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:686, code: return dt.redistribute(mesh, [Replicate()]).to_local() + 2
        prim_redistribute = torch__dynamo_variables_tensor_prim_redistribute(prim_from_local);  prim_from_local = None
        to_local = prim_redistribute.to_local();  prim_redistribute = None
        add = to_local + 2;  to_local = None
        return (add,)
```
since the `args` here are "device_mesh" and "placements" only (`redistribute` have a fixed signature), so the `args_as_value` would be safely encoded inside the `redistribute_fn_with_prim_types` and we insert a call_function node inside the graph.

voznesenskym replied:
As long as this runs correctly, I am fine. I was skeptical, but its nice to see.

wanchaol commented on torch/_dynamo/variables/builder.py:
```diff
@@ -634,6 +627,26 @@ def index_source(key):
+            return DeviceMeshVariable(
+                value,
+                source=self.source,
+                guards=self.make_guards(GuardBuilder.ID_MATCH),
```
good point, feels like this should be a type match instead, I found ProcessGroupVariable uses ID_MATCH so I copied it, maybe we should change that to type match too

voznesenskym replied:
Understood, you are using the contents of the object. 

ID_MATCH is good, lets leave a TODO to give it its own guard type at some point, since ostensibly DeviceMesh created with the same inputs, but different id, should be ok.

wanchaol replied:
sounds good! done.

wanchaol committed:
Update on "[dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it:
- allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.
- We use `allow_in_graph` to include `DTensor` and `DTensor.from_local` to be represented as `TorchVariable`
- The dtensor created becomes a normal `TensorVariable` and it would insert any tensor operations to the output graph just like torch.Tensor
- note that dtensor have a new instance method `redistribute` compare to plain tensor, and we currently special handle it in `TensorVariable`

`from_local` and `redistribute` both accepts some non-trival metadata as arguments (i.e. DeviceMesh, Placement) which fx.Graph does not support. In order to let these two APIs appear in the dynamo captured graph, we encoded the metadata into a new_function (like `functools.partial`) and the new function only accepts prim args (i.e. tensor), then we put `call_function` with this new_function to the graph. This is suggested by ezyang. The underlying rationale here is that the metadata will not change across the graph invocations so it's safe to encode them.


Captured graph:
```
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:685, code: dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
        prim_from_local = torch__dynamo_variables_torch_prim_from_local(l_x_, run_check = False);  l_x_ = None
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:686, code: return dt.redistribute(mesh, [Replicate()]).to_local() + 2
        prim_redistribute = torch__dynamo_variables_tensor_prim_redistribute(prim_from_local);  prim_from_local = None
        to_local = prim_redistribute.to_local();  prim_redistribute = None
        add = to_local + 2;  to_local = None
        return (add,)
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

voznesenskym reviewed (approved):
LGTM, ty for addressing nits. The graph you responded with looks reasonable.

wanchaol committed:
Update on "[dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it:
- allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.
- We use `allow_in_graph` to include `DTensor` and `DTensor.from_local` to be represented as `TorchVariable`
- The dtensor created becomes a normal `TensorVariable` and it would insert any tensor operations to the output graph just like torch.Tensor
- note that dtensor have a new instance method `redistribute` compare to plain tensor, and we currently special handle it in `TensorVariable`

`from_local` and `redistribute` both accepts some non-trival metadata as arguments (i.e. DeviceMesh, Placement) which fx.Graph does not support. In order to let these two APIs appear in the dynamo captured graph, we encoded the metadata into a new_function (like `functools.partial`) and the new function only accepts prim args (i.e. tensor), then we put `call_function` with this new_function to the graph. This is suggested by ezyang. The underlying rationale here is that the metadata will not change across the graph invocations so it's safe to encode them.


Captured graph:
```
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:685, code: dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
        prim_from_local = torch__dynamo_variables_torch_prim_from_local(l_x_, run_check = False);  l_x_ = None
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:686, code: return dt.redistribute(mesh, [Replicate()]).to_local() + 2
        prim_redistribute = torch__dynamo_variables_tensor_prim_redistribute(prim_from_local);  prim_from_local = None
        to_local = prim_redistribute.to_local();  prim_redistribute = None
        add = to_local + 2;  to_local = None
        return (add,)
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

wanchaol committed:
Update on "[dynamo] add initial dynamo support for DTensor"


This PR adds initial dynamo support for DTensor, in particular, it:
- allows DTensor be passed into a compiled function, and allow fakify
DTensor during dynamo tracing by turning the inner local tensor to meta
tensor.
- We use `allow_in_graph` to include `DTensor` and `DTensor.from_local` to be represented as `TorchVariable`
- The dtensor created becomes a normal `TensorVariable` and it would insert any tensor operations to the output graph just like torch.Tensor
- note that dtensor have a new instance method `redistribute` compare to plain tensor, and we currently special handle it in `TensorVariable`

`from_local` and `redistribute` both accepts some non-trival metadata as arguments (i.e. DeviceMesh, Placement) which fx.Graph does not support. In order to let these two APIs appear in the dynamo captured graph, we encoded the metadata into a new_function (like `functools.partial`) and the new function only accepts prim args (i.e. tensor), then we put `call_function` with this new_function to the graph. This is suggested by ezyang. The underlying rationale here is that the metadata will not change across the graph invocations so it's safe to encode them.


Captured graph:
```
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:685, code: dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)
        prim_from_local = torch__dynamo_variables_torch_prim_from_local(l_x_, run_check = False);  l_x_ = None
        
        # File: /scratch/wanchaol/work/pytorch/test/distributed/_tensor/test_dtensor.py:686, code: return dt.redistribute(mesh, [Replicate()]).to_local() + 2
        prim_redistribute = torch__dynamo_variables_tensor_prim_redistribute(prim_from_local);  prim_from_local = None
        to_local = prim_redistribute.to_local();  prim_redistribute = None
        add = to_local + 2;  to_local = None
        return (add,)
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
